Archivist.lua

```diff
@@ -68,39 +67,146 @@ function Archivist:Warn(valid, pattern, ...) -- Like assert, but doesn't interru
+	local function randomHex()
+		return ('%x'):format(math.random(0, 0xf))
+	end
+
+	function proto:GenerateID()
+		local template ='xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx'
+		return (template:gsub('x', randomHex))
+	end
+end
+
+function Archivist:IsInitialized(sv)
+	return self.archives[sv] ~= nil
 end
 
--- Give Archivist its archive to play with. Called automatically, unless Archivist has been embedded.
-function Archivist:Initialize(sv)
+-- Create an archivist instance. Called automatically, unless Archivist has been embedded.
+-- If called with an already initialized savedvariables table, then the previously initialized instance is returned.
+function Archivist:Initialize(sv, prototypes)
 	do -- arg validation
-		self:Assert(not self:IsInitialized(), "Archivist has already been initialized.")
-		self:Assert(type(sv) == "table", "Attempt to initialize Archivist SavedVariables with a %q instead of a table.", type(sv))
+		self:Assert(type(sv) == "table", "Invalid argument #1 to Initialize, expected table but got %q.", type(sv))
+		self:Assert(type(sv.internalVersion) ~= "number" or sv.internalVersion <= self.internalVersion, "Invalid argument #1 to Initialize, savedvariable is encoded using a newer version of Archivist. Please upgrade this installation if possible.")
+		self:Assert(prototypes == nil or type(prototypes) == "table", "Invalid argument #2 to Initialize, expected table or nil but got %q", type(prototypes))
+	end
+	if self:IsInitialized(sv) then
+		return self.archives[sv]
+	end
+	if type(sv.stores) ~= "table" then
+		sv.stores = {}
+	end
+	local archive = setmetatable({
+		prototypes = {},
+		activeStores = {},
+		storeMap = {},
+		sv = sv,
+		initialized = true,
+	}, {
+		__index = proto
+	})
+	if type(sv.internalVersion) ~= "number" or sv.internalVersion < self.internalVersion then
+		self:DoMigration(archive)
+	end
+	self.archives[sv] = archive
+	for _, prototype in pairs(self.defaultStoreTypes) do
+		archive:RegisterStoreType(prototype)
+	end
+	if prototypes then
+		for _, prototype in pairs(prototypes) do
+			archive:RegisterStoreType(prototype)
+		end
 	end
+	return archive
+end
+
+-- register a migration to upgrade archivist version
+function Archivist:RegisterMigration(version, migration)
+	do -- arg validation
+		self:Assert(type(version) == "number" and version <= self.internalVersion and version % 1 == 0 and self.migrations[version] == nil,
+			"Migration version should be valid integer")
+		self:Assert(type(migration) == "function", "Migration should be a function")
+	end
+	self.migrations[version] = migration
+end
 
-	self.sv = sv
-	self.initialized = true
-	for id, prototype in pairs(self.prototypes) do
-		self.sv[id] = self.sv[id] or {}
-		if prototype.Init then
-			prototype:Init()
+function Archivist:DoMigration(archive)
+	for i = archive.sv.internalVersion or 1, self.internalVersion do
+		if self.migrations[i] then
+			self.migrations[i](archive)
 		end
 	end
+	archive.sv.internalVersion = self.internalVersion
```

Comment:
```suggestion
		if self.migrations[i] then
			self.migrations[i](archive)
		end

		archive.sv.internalVersion = i
	end
```

This is mostly a safety suggestion; if multiple migrations ever exist and an error occurs in anything beyond the first applied one then that error would prevent the internal version from being updated. This would lead to a situation where we'd possibly re-apply migrations that already took place on the next initialization of the same data, which could cause more problems.